Work analytics

gordonleacock opened the pull request:
Issue #1455, work in progress, no spec tests yet.

njaffer commented:
@gordonleacock , your two branches (for tags and work analytics) are both combined in this PR. You might need to remove the changes for the 'tags' from this PR.

njaffer commented:
The code for app/views/curation_concerns/base/_show_actions.html.erb should work as is, you do not need to change it, the path was changed in the PR for file analytics here: https://github.com/projecthydra/sufia/pull/1540/files

gordonleacock commented:
Removed the coll_tag_fix changes and reverted app/views/curation_concerns/base/_show_actions.html.erb back to master version. Also (hopefully) I fixed a problem on L96 of user_stat_importer.rb. Will issue another pull request soon.

jcoyne commented on .gitignore:
```diff
@@ -6,6 +6,9 @@ ruby-gemset
 Gemfile.lock
 
+# Ignore Fedora data files
+/fcrepo4-data
```
Duplicate of line 13?

gordonleacock replied:
Thanks @jcoyne, yes and I'll fix that.

jcoyne commented:
Is there a reason we're adding work_view_stats and not reusing file_view_stats?  There is really no need to distinguish between the view of a work and the view of a file.  Given the id of each, we can identify if the node is a file or a work.

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
 
     private
 
-      def delay
```
Why are you changing the indentation here?  Please revert this. We have been using the style where you "Indent after private/protected" http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions

gordonleacock replied:
Oops - didn't mean to do that.

jcoyne commented on app/views/shared/_citations.html.erb:
```diff
@@ -15,18 +15,18 @@
-    <meta name="citation_firstpage" content=""/>
-    <meta name="citation_lastpage" content=""/>
-  -->
+    <meta name="citation_title" content="<%= @presenter.title.first %>"/>
```
Please use 2-space indentation style per http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions

gordonleacock replied:
@jcoyne - Roger, will fix!

jcoyne commented on app/presenters/sufia/collection_presenter.rb:
```diff
@@ -8,7 +8,7 @@ class CollectionPresenter < CurationConcerns::CollectionPresenter
     def self.terms
       [:title, :total_items, :size, :resource_type, :description, :creator,
-       :contributor, :tag, :rights, :publisher, :date_created, :subject,
+       :contributor, :tags, :rights, :publisher, :date_created, :subject,
```
Why is this being changed? Is  this is relevant to the purpose of this pull request?

jcoyne commented on app/models/work_view_stat.rb:
```diff
@@ -0,0 +1,19 @@
+class WorkViewStat < ActiveRecord::Base
+  extend Sufia::WorkStatUtils
```
Why are you putting these methods in a module?  It doesn't appear that the module would be able to be used in any other context.

gordonleacock replied:
For work statistics I've been following the pattern already in place for files statistics.

mjgiarlo replied:
:+1:

jcoyne commented on spec/presenters/sufia/collection_presenter_spec.rb:
```diff
@@ -4,7 +4,7 @@
     subject { described_class.terms }
     it { is_expected.to eq [:title, :total_items, :size, :resource_type, :description, :creator,
-                            :contributor, :tag, :rights, :publisher, :date_created, :subject,
+                            :contributor, :tags, :rights, :publisher, :date_created, :subject,
```
Why is this being changed?

jcoyne commented on app/views/curation_concerns/base/_show_actions.html.erb:
```diff
@@ -1,30 +1,37 @@
 <div class="form-actions">
   <% if Sufia.config.analytics %>
-    <%= link_to "Analytics", @presenter.stats_path, id: 'stats', class: 'btn btn-default' %>
+      <% if @presenter.model_name.name ==  "GenericWork" %>
```
Please use 2 space indentation.

jcoyne commented on app/models/work_usage.rb:
```diff
@@ -0,0 +1,44 @@
+class WorkUsage
```
Please add class level documentation.  What does this do?

mjgiarlo replied:
@gordonleacock is following the pattern already used by Sufia in `app/models/file_usage.rb`. I suggest we hold off on any refactors (of moving model stuff into presenters) until this PR is merged.

jcoyne commented on app/models/work_usage.rb:
```diff
@@ -0,0 +1,44 @@
+  # work.create_date reflects the date the work was added to Fedora. On data
+  # migrated from one repository to another the created_date can be later
+  # than the date the work was uploaded.
+  def date_for_analytics(work)
```
Should this be a private method?

jcoyne commented on app/models/work_usage.rb:
```diff
@@ -0,0 +1,44 @@
+    earliest > date_analytics ? earliest : date_analytics
+  end
+
+  def string_to_date(date_str)
```
make this private?

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -28,6 +29,13 @@ def import
           delay
         end
 
+        # Added 1-26-16 gordonl
```
No need to comment this, git will track it for us.

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
-          total_stats[date_key].store(stat_name, new_count)
-        end
-        total_stats
+    # Added 1-26-16 gordonl
```
No need to comment this, git will track it for us.

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
-      def create_or_update_user_stats(stats, user)
-        stats.each do |date_string, data|
-          date = Time.zone.parse(date_string)
+    # Modified 1-26-16 gordonl
```
No need to comment this. Git will track it.

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
-          user_stat.save!
-        end
-      end
+        user_stat = UserStat.where(user_id: user.id).where(date: date).first
```
You can combine the where clauses:  `UserStat.where(user_id: user.id, date: date)...`

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
-        end
-      end
+        user_stat = UserStat.where(user_id: user.id).where(date: date).first
+        user_stat ||= UserStat.new(user_id: user.id, date: date)
```
Instead of doing this, can you do `user_stat = UserStat.where(...).first_or_initialize` on the previous line.

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
-      def log_message(message)
-        puts message if @verbose
-        Rails.logger.info "#{self.class}: #{message}" if @logging
+        user_stat.file_views = data[:views] || 0
```
how about `data.fetch(:views, 0)`

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
+    end
+
+    def log_message(message)
+      puts message if @verbose
```
Please don't puts log messages. If you want them to go to stdout, set the `Rails.logger`

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -46,81 +54,93 @@ def sorted_users
+          delay
+          retry
+        else
+          log_message fail_message
```
This message should be logged at an error level, not the info level that is done by default.

mjgiarlo replied:
@jcoyne: I believe all that @gordonleacock has changed here is the indentation. We should probably handle this in a separate PR.

jcoyne commented on app/models/concerns/sufia/work_stat_utils.rb:
```diff
@@ -0,0 +1,34 @@
+
+    private
+
+    def cached_stats(work_id, start_date, _method)
```
Indent all methods after private/protected per the rails style: http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions

jcoyne commented on app/models/concerns/sufia/work_stat_utils.rb:
```diff
@@ -0,0 +1,34 @@
+    end
+  end
+end
+``
```
Remove this line

jcoyne commented on app/models/work_view_stat.rb:
```diff
@@ -0,0 +1,19 @@
+class WorkViewStat < ActiveRecord::Base
```
Add a unit test for this class

gordonleacock replied:
Added spec/models/work_view_stat_spec.rb

mjgiarlo replied:
:+1:

jcoyne commented:
This change seems to have broken the spec/lib/sufia/user_stat_importer_spec.rb tests

jcoyne commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -155,6 +196,7 @@
           expect(UserStat.count).to eq 8 # 2 users for 4 days
           UserStat.all.each do |stat|
             expect(stat.file_views).to eq 0
+            #expect(stat.work_views).to eq 0
```
why are you removing this test?

mjgiarlo replied:
Uncomment, pls.

jcoyne commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -169,6 +211,7 @@
           expect(UserStat.count).to eq 8 # 2 users for 4 days
           UserStat.all.each do |stat|
             expect(stat.file_views).not_to eq 0
+            #expect(stat.work_views).not_to eq 0
```
why is this test removed?

mjgiarlo replied:
Uncomment, pls.

gordonleacock replied:
Ack, reverting problem again. Guess it is better to do a diff so I don't over-revert!

jcoyne commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -137,6 +177,7 @@
       context "both error out completely" do
         before do
           expect(FileDownloadStat).to receive(:ga_statistics).exactly(12).times.and_raise(StandardError.new("GA error"))
+          #expect(WorkViewStat).to receive(:ga_statistics).exactly(12).times.and_raise(StandardError.new("GA error"))
```
Is this needed or not?

gordonleacock replied:
Still sorting out the spec. Those were items I'm still working on.

On Wed, Feb 17, 2016 at 12:31 PM, Justin Coyne <notifications@github.com>
wrote:

> In spec/lib/sufia/user_stat_importer_spec.rb
> <https://github.com/projecthydra/sufia/pull/1564#discussion_r53199877>:
>
> > @@ -137,6 +177,7 @@
> >        context "both error out completely" do
> >          before do
> >            expect(FileDownloadStat).to receive(:ga_statistics).exactly(12).times.and_raise(StandardError.new("GA error"))
> > +          #expect(WorkViewStat).to receive(:ga_statistics).exactly(12).times.and_raise(StandardError.new("GA error"))
>
> Is this needed or not?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/projecthydra/sufia/pull/1564/files#r53199877>.
>

mjgiarlo replied:
Uncomment, pls.

mjgiarlo commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -31,6 +31,13 @@
   end
 
+  # work
+  let!(:bilbo_work_1) { build(:generic_work, id: 'xyzbilbowork1', title: ['bilbo work 1'], depositor: :bilbo) }
```
Try this syntax instead:

```ruby
let!(bilbo_work_1) do
  GenericWork.create(id: 'xyzbilbowork1', title: ['bilbo work 1']) do |work|
    work.apply_depositor_metadata(bilbo.email)
  end
end
```

gordonleacock commented:
2-18-16 UPDATE:
Still ugly - not passing all specs. Would welcome suggestions.

mjgiarlo commented on app/models/work_usage.rb:
```diff
@@ -0,0 +1,50 @@
+# class WorkUsage follows the model established by FileUsage
+# Called by the stats controller, it finds cached work pageview data,
+# and prepares it for visualization in /app/views/stats/work.html.erb
+
```
You can remove this space, but otherwise I believe you've satisfied @jcoyne's request below. :+1:

mjgiarlo commented on lib/generators/sufia/cached_stats_generator.rb:
```diff
@@ -16,7 +16,8 @@ def banner
       'create_file_view_stats.rb',
-      'create_file_download_stats.rb'
+      'create_file_download_stats.rb',
+      'create_work_view_stats.rb'
```
We'll want this in a new generator (`cached_work_stats_generator.rb`?) so that folks who already have the other two migrations can pick the new one up independent of the old ones.

jcoyne replied:
Since the URLs for files/works are changing, do we need to keep the sufia 6 statistics in their own table?

mjgiarlo commented on lib/generators/sufia/templates/migrations/create_user_stats.rb:
```diff
@@ -5,15 +5,18 @@ def change
       t.datetime :date
       t.integer :file_views
       t.integer :file_downloads
+      t.integer :work_views
```
We'll want this in a new migration that your new generator (re: :arrow_heading_up:) can include.

mjgiarlo commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -234,23 +289,34 @@
-      bilbo_file_2_pageview_stats
-    else
-      frodo_file_1_pageview_stats
+      when bilbo_file_1.id
```
Revert these indentation changes, pls.

mjgiarlo commented on spec/lib/sufia/user_stat_importer_spec.rb:
```diff
@@ -46,61 +65,92 @@
-      OpenStruct.new(eventCategory: "Files", eventAction: "Downloaded", eventLabel: "bilbo1", date: date_strs[2], totalEvents: "5"),
-      OpenStruct.new(eventCategory: "Files", eventAction: "Downloaded", eventLabel: "bilbo1", date: date_strs[3], totalEvents: "3"),
-      OpenStruct.new(eventCategory: "Files", eventAction: "Downloaded", eventLabel: "bilbo1", date: date_strs[4], totalEvents: "7")
+        OpenStruct.new(eventCategory: "Files", eventAction: "Downloaded", eventLabel: "bilbo1", date: date_strs[0], totalEvents: "2"),
```
Why all the indentation changes in this file?

gordonleacock replied:
Sorry. Ran into problems then reverted but forgot to fix indents again. Will fix before next PR.

mjgiarlo commented on spec/models/work_usage_spec.rb:
```diff
@@ -0,0 +1,165 @@
+
+describe WorkUsage, type: :model do
+  let(:work) do
+    GenericWork.new.tap do |work|
```
If you use this syntax instead, you can save a line and the call to `tap`:

```ruby
# may also need to pass a title arg to #create here, but maybe not
GenericWork.create do |work|
  work.apply_depositor_metadata('awead')
end
```

mjgiarlo commented on spec/models/work_view_stat_spec.rb:
```diff
@@ -0,0 +1,82 @@
+        OpenStruct.new(date: date_strs[1], pageviews: 8),
+        OpenStruct.new(date: date_strs[2], pageviews: 6),
+        OpenStruct.new(date: date_strs[3], pageviews: 10),
+        # OpenStruct.new(date: date_strs[4], pageviews: 2)
```
Why's this one commented out? Was it also commented in the file you used as a pattern for this?

gordonleacock replied:
Yes it was commented out in the file_view_stat_spec.rb that I used as a model.

mjgiarlo replied:
OK!

jcoyne commented on app/services/sufia/user_stat_importer.rb:
```diff
@@ -84,13 +89,19 @@ def file_ids_for_user(user)
-      #  "2014-12-01 00:00:00 UTC" => {:views=>4, :downloads=>4}}
-      def tally_results(file_stats, stat_name, total_stats)
-        file_stats.each do |stats|
+      def work_ids_for_user(user)
```
Should this move to the User model?  `user.work_ids`?

mjgiarlo replied:
It could and probably should. If it does, so should `file_ids_for_user`, so I'm inclined to say that should be a separate issue and PR.

jcoyne replied:
The reason I'm suggesting this here is it may allow us to avoid having a ClassLength exception in rubocop_todo.yml

mjgiarlo replied:
¯\_(ツ)_/¯ I'm inclined to get this PR merged before @gordonleacock needs to rebase it again.

mjgiarlo commented:
Fixes #1455.
